### PR TITLE
feat: add singleflight to transport

### PR DESCRIPTION
**Pull request by jensneuse**

### CODE REVIEW

**Review by Aenimus — APPROVED** [review]

File: router/core/transport.go [hunk 1]
```diff
@@ -78,6 +89,107 @@ func (ct *CustomTransport) RoundTrip(req *http.Request) (*http.Response, error)
...
+	if err != nil {
+		return nil, err
+	}
+
+	sfStats := resolve.GetSingleFlightStats(req.Context())
+	if sfStats != nil {
+		sfStats.SingleFlightUsed = true
+		sfStats.SingleFlightSharedResponse = shared
+	}
+	rwb := v.(*responseWithBody)
```
Comment (StarpTech) [inline router/core/transport.go:177]: Do a conversion check here.

File: router/core/transport.go [hunk 2]
```diff
@@ -78,6 +89,107 @@ func (ct *CustomTransport) RoundTrip(req *http.Request) (*http.Response, error)
...
+	if req.Body != nil {
+		body, err := io.ReadAll(req.Body)
+		if err != nil {
+			return nil, err
+		}
+		_, _ = keyGen.Write(body)
+		req.Body = io.NopCloser(bytes.NewReader(body))
+	}
+
+	unsortedHeaders := make([]string, 0, len(req.Header))
```
Comment (StarpTech) [inline router/core/transport.go:139]: Can we use `req.Headers.Clone()` here?

File: router/core/transport.go [hunk 3]
```diff
@@ -78,6 +89,107 @@ func (ct *CustomTransport) RoundTrip(req *http.Request) (*http.Response, error)
...
+		return nil, err
+	}
+
+	sfStats := resolve.GetSingleFlightStats(req.Context())
+	if sfStats != nil {
+		sfStats.SingleFlightUsed = true
+		sfStats.SingleFlightSharedResponse = shared
+	}
+	rwb := v.(*responseWithBody)
+	res := &http.Response{}
```
Comment (StarpTech) [inline router/core/transport.go:178]: Some props are missing like proto*, close, TLS Check the struct definition on core.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: router/core/transport.go (hunk 1 above)
Comment (jensneuse) [reply router/core/transport.go:177]: it's impossible

File: router/core/transport.go (hunk 2 above)
Comment (jensneuse) [reply router/core/transport.go:139]: no need to

File: router/core/transport.go [hunk 4]
```diff
@@ -78,6 +89,107 @@ func (ct *CustomTransport) RoundTrip(req *http.Request) (*http.Response, error)
...
+
+	keyGen := pool.Hash64.Get()
+	defer pool.Hash64.Put(keyGen)
+
+	if req.Body != nil {
+		body, err := io.ReadAll(req.Body)
+		if err != nil {
+			return nil, err
+		}
+		_, _ = keyGen.Write(body)
```
[the remark this answers is not in the document]
Comment (jensneuse) [reply router/core/transport.go]: done

File: router/core/transport.go [hunk 5]
```diff
@@ -78,6 +89,107 @@ func (ct *CustomTransport) RoundTrip(req *http.Request) (*http.Response, error)
...
+	unsortedHeaders := make([]string, 0, len(req.Header))
+
+	for key := range req.Header {
+		value := req.Header.Get(key)
+		unsortedHeaders = append(unsortedHeaders, key+value)
+	}
+
+	sort.Strings(unsortedHeaders)
+	for i := range unsortedHeaders {
+		_, _ = keyGen.Write(unsafebytes.StringToBytes(unsortedHeaders[i]))
```
[the remark this answers is not in the document]
Comment (jensneuse) [reply router/core/transport.go]: done

**Review by StarpTech — APPROVED** [review]
LGTM

**Pull request merged**